adapters/silverpush/silverpush.go

```diff
@@ -0,0 +1,346 @@
+	imps := request.Imp
+	requests := make([]*adapters.RequestData, 0, len(imps))
+	errors := make([]error, 0, len(imps))
+
+	for _, imp := range imps {
+		impsByMediaType, err := impressionByMediaType(&imp)
+		if err != nil {
+			errors = append(errors, err)
+		}
+
+		for _, impByMediaType := range impsByMediaType {
+			request.Imp = []openrtb2.Imp{impByMediaType}
+
+			if err := validateRequest(request); err != nil {
+				errors = append(errors, err)
+				continue
+			}
+			requestData, err := a.makeRequest(request)
+			if err != nil {
+				errors = append(errors, err)
+				continue
+			}
+
+			requests = append(requests, requestData)
+
+		}
+
+	}
+
+	return requests, errors
+
+}
+
+func (a *adapter) makeRequest(req *openrtb2.BidRequest) (*adapters.RequestData, error) {
+	reqJSON, err := json.Marshal(req)
+	if err != nil {
+		return nil, err
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+	headers.Add("X-Openrtb-Version", "2.5")
+	return &adapters.RequestData{
+		Method:  "POST",
+		Uri:     a.endpoint,
+		Body:    reqJSON,
+		Headers: headers,
+	}, nil
+}
+
+func validateRequest(req *openrtb2.BidRequest) error {
+	imp := &req.Imp[0]
+	var silverPushExt openrtb_ext.ImpExtSilverpush
+	err := setPublisherId(req, imp, &silverPushExt)
+	if err != nil {
+		return err
+	}
+	if err := setUser(req); err != nil {
+		return err
+	}
+	setDevice(req)
+	setExtToRequest(req, silverPushExt.PublisherId)
+	return setImpForAdExchange(imp, &silverPushExt)
+}
+
+func setDevice(req *openrtb2.BidRequest) {
+	if req.Device != nil {
+		deviceCopy := *req.Device
+		if len(deviceCopy.UA) > 0 {
+			deviceCopy.OS = getOS(deviceCopy.UA)
+			if isMobile(deviceCopy.UA) {
+				deviceCopy.DeviceType = 1
+			} else if isCTV(deviceCopy.UA) {
+				deviceCopy.DeviceType = 3
+			} else {
+				deviceCopy.DeviceType = 2
+			}
+		}
+		req.Device = &deviceCopy
+	}
+}
+
+func setUser(req *openrtb2.BidRequest) error {
+	var extUser openrtb_ext.ExtUser
+	var userExtRaw map[string]json.RawMessage
+
+	if req.User != nil && req.User.Ext != nil {
+		if err := json.Unmarshal(req.User.Ext, &userExtRaw); err != nil {
+			return &errortypes.BadInput{Message: "Invalid user.ext."}
+		}
+		if userExtDataRaw, ok := userExtRaw["data"]; ok {
+			if err := json.Unmarshal(userExtDataRaw, &extUser); err != nil {
+				return &errortypes.BadInput{Message: "Invalid user.ext.data."}
+			}
+			var userCopy = *req.User
+			if IsValidEids(extUser.Eids) {
+				userExt, _ := json.Marshal(
+					&openrtb2.User{
+						EIDs: extUser.Eids,
+					})
+				userCopy.Ext = userExt
+				req.User = &userCopy
+			}
+		}
+	}
+	return nil
+}
+
+func setExtToRequest(req *openrtb2.BidRequest, publisherID string) {
+
+	record := map[string]string{
+		"bc":          bidderConfig + "_" + bidderVersion,
+		"publisherId": publisherID,
+	}
+	reqExt, _ := json.Marshal(record)
```

Comment:
Please do not ignore errors.

Reply:
Fixed
[commit](https://github.com/prebid/prebid-server/pull/2957/commits/cc30407f798a60b4550738f3802ffe7a19f81493)